File: stats.py
# -*- coding: utf-8 -*-
"""
实时统计模块
- CPM (每分钟按键数) 计算，带锁 + 缓存
- 跨天自动重置
"""

import time
import threading
from collections import deque
from typing import Optional

from config import config
from logger import get_logger

log = get_logger('stats')


class CPMCalculator:
    """实时打字速度计算器

    优化点：
    - deque + 锁，线程安全
    - 查询时惰性清理过期数据
    - 结果缓存（避免每次查询都遍历）
    """

    def __init__(self, window: int):
        self.window = window
        self._timestamps: deque = deque()
        self._lock = threading.Lock()
        self._cached_cpm = 0
        self._cached_at = 0.0
        self._cache_ttl = 0.5  # 500ms 缓存

    def record(self):
        """记录一次按键时间戳"""
        now = time.time()
        with self._lock:
            self._timestamps.append(now)
            # 顺便清理一下头部过期数据，控制 deque 长度
            cutoff = now - self.window
            while self._timestamps and self._timestamps[0] < cutoff:
                self._timestamps.popleft()
            # 写入时缓存失效
            self._cached_at = 0.0

    def get_cpm(self) -> int:
        """获取当前 CPM"""
        # 命中缓存直接返回
        now = time.time()
        if now - self._cached_at < self._cache_ttl:
            return self._cached_cpm

        cutoff = now - self.window
        with self._lock:
            while self._timestamps and self._timestamps[0] < cutoff:
                self._timestamps.popleft()
            count = len(self._timestamps)
        # 缓存结果
        self._cached_cpm = count
        self._cached_at = now
        return count

    def reset(self):
        with self._lock:
            self._timestamps.clear()
            self._cached_at = 0.0


# 全局实例（延迟创建，便于配置变更）
_cpm: Optional[CPMCalculator] = None
_cpm_lock = threading.Lock()


def get_cpm_calculator() -> CPMCalculator:
    global _cpm
    if _cpm is None:
        with _cpm_lock:
            if _cpm is None:
                window = config.getint('stats', 'cpm_window', 60)
                _cpm = CPMCalculator(window)
                log.info('CPM 计算器初始化 (window=%ds)', window)
    return _cpm


def record_cpm():
    get_cpm_calculator().record()


def get_current_cpm() -> int:
    return get_cpm_calculator().get_cpm()


def reset_cpm():
    """重置 CPM 计数器（清除当前按键时间戳）

    用于清除某按键今日次数后，立即清掉可能虚高的实时速度。
    """
    try:
        get_cpm_calculator().reset()
    except Exception as e:
        log.debug('重置 CPM 失败: %s', e)

File: crates/core/src/stats.rs
//! 实时统计：CPM（每分钟操作数）计算。
//!
//! 镜像 Python 版 `stats.py`：
//! - 滑动时间窗口 + deque 上限安全阀
//! - 写入/查询分离，查询时惰性清理过期数据
//! - 结果缓存（TTL 500ms）
//! - 线程安全

use std::collections::VecDeque;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use crate::config::FocusFlowConfig;

/// CPM 内部状态（时间戳队列 + 结果缓存合并为单锁，减少热路径锁争用）。
struct CpmState {
    /// 时间戳队列（单调时钟 Instant）
    timestamps: VecDeque<Instant>,
    /// 缓存结果
    cached_count: i64,
    cached_at: Instant,
}

/// CPM 计算器。
pub struct CpmCalculator {
    /// 窗口（秒）
    window: f64,
    state: Mutex<CpmState>,
}

impl CpmCalculator {
    pub fn new(window: f64) -> Self {
        Self {
            window: window.max(1.0),
            state: Mutex::new(CpmState {
                timestamps: VecDeque::with_capacity(4096),
                cached_count: 0,
                cached_at: Instant::now() - std::time::Duration::from_secs(10),
            }),
        }
    }

    /// 记录一次操作时间戳。
    pub fn record(&self) {
        let now = Instant::now();
        let mut state = self.state.lock().unwrap();
        state.timestamps.push_back(now);
        // 硬上限安全阀（对应 Python maxlen=100000）
        if state.timestamps.len() > 100_000 {
            state.timestamps.pop_front();
        }
        // 顺带清理窗口外旧数据
        let cutoff = now - std::time::Duration::from_secs_f64(self.window);
        while let Some(&front) = state.timestamps.front() {
            if front < cutoff {
                state.timestamps.pop_front();
            } else {
                break;
            }
        }
        // 写入时缓存失效
        state.cached_count = 0;
        state.cached_at = now - std::time::Duration::from_secs(10);
    }

    /// 获取当前 CPM（窗口内事件数）。
    pub fn get_cpm(&self) -> i64 {
        let now = Instant::now();
        let mut state = self.state.lock().unwrap();
        if state.cached_at.elapsed() < std::time::Duration::from_millis(500) {
            return state.cached_count;
        }
        let cutoff = now - std::time::Duration::from_secs_f64(self.window);
        while let Some(&front) = state.timestamps.front() {
            if front < cutoff {
                state.timestamps.pop_front();
            } else {
                break;
            }
        }
        let count = state.timestamps.len() as i64;
        state.cached_count = count;
        state.cached_at = now;
        count
    }

    /// 重置（清除当前时间戳）。
    pub fn reset(&self) {
        let mut state = self.state.lock().unwrap();
        state.timestamps.clear();
        state.cached_count = 0;
        state.cached_at = Instant::now() - std::time::Duration::from_secs(10);
    }
}

/// 全局 CPM 单例。
static CPM: std::sync::OnceLock<Arc<CpmCalculator>> = std::sync::OnceLock::new();

/// 获取全局 CPM 计算器。
pub fn cpm(config: &'static FocusFlowConfig) -> Arc<CpmCalculator> {
    Arc::clone(CPM.get_or_init(|| {
        let window = config.get_float("stats", "cpm_window", 60.0);
        Arc::new(CpmCalculator::new(window))
    }))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cpm_window() {
        let calc = CpmCalculator::new(60.0);
        for _ in 0..10 {
            calc.record();
        }
        assert_eq!(calc.get_cpm(), 10);
        calc.reset();
        assert_eq!(calc.get_cpm(), 0);
    }
}
